### ISSUE TITLE
sqlite user-defined functions fail with OperationalError

### ISSUE THREAD

**Issue by commonism** [sqlalchemy/sqlalchemy#5623]
**Describe the bug**
Kown to work code fails to run using SQLAlchemy-1.4.0b1.dev0.

**Expected behavior**
The user-defined function is supposed to work.

**To Reproduce**

```py
# sa_sqlite3_regex.py - use Python re for sqlite3 REGEXP operator w/ sqlalchemy

import re

import sqlalchemy as sa
import sqlalchemy.ext.declarative

@sa.event.listens_for(sa.engine.Engine, 'connect')
def sqlite_engine_connect(dbapi_conn, connection_record):
    dbapi_conn.create_function('regexp', 2, _regexp)

def _regexp(pattern, value):
    if value is None:
        return None
    return re.search(pattern, value) is not None

class Spam(sa.ext.declarative.declarative_base()):

    __tablename__ = 'spam'

    id = sa.Column(sa.Integer, primary_key=True)
    name = sa.Column(sa.Text)

engine = sa.create_engine('sqlite://', echo=True)

Spam.metadata.create_all(engine)

sa.insert(Spam, bind=engine).execute([
    {'name': 'spam'},
    {'name': 'eggs'},
    {'name': 'ham'},
    {'name': None},
])

query = sa.select([Spam], bind=engine).where(Spam.name.op('REGEXP')(r'(\w)\1'))

print(query.execute().fetchall())
```
(found here: https://gist.githubusercontent.com/xflr6/c7c87befd9e88e8b1fcb0f042610eea4/raw/77506058a8c0f659fd7bd67e2bf7de23ae2c96c4/sa_sqlite3_regex.py)

**Error**

```
Traceback (most recent call last):
  File "~/venv-le/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1617, in _execute_context
    self.dialect.do_execute(
  File "~/venv-le/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 662, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: user-defined function raised exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "tests/sqlite_regex.py", line 45, in <module>
    print(query.execute().fetchall())
  File "<string>", line 2, in execute
  File "~/venv-le/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py", line 347, in warned
    return fn(*args, **kwargs)
  File "~/venv-le/lib/python3.8/site-packages/sqlalchemy/sql/base.py", line 885, in execute
    return e._execute_clauseelement(
  File "~/venv-le/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 2889, in _execute_clauseelement
    return connection._execute_clauseelement(
  File "~/venv-le/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1304, in _execute_clauseelement
    ret = self._execute_context(
  File "~/venv-le/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1660, in _execute_context
    self._handle_dbapi_exception(
  File "~/venv-le/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1841, in _handle_dbapi_exception
    util.raise_(
  File "~/venv-le/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 180, in raise_
    raise exception
  File "~/venv-le/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1617, in _execute_context
    self.dialect.do_execute(
  File "~/venv-le/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 662, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) user-defined function raised exception
[SQL: SELECT spam.id, spam.name 
FROM spam 
WHERE spam.name regexp ?]
[parameters: ('(\\w)\\1',)]
```

** Expected **
```
[(2, 'eggs')]
```

**Versions.**
 - OS: Linux
 - Python: 3.8
 - SQLAlchemy: SQLAlchemy-1.4.0b1.dev0
 - Database: sqlite
 - DBAPI: sqlite3

**Comment by zzzeek** [comment]
hi -

there's a bug here but it's not what you're reporting, that will be fixed in #5624.

your custom regex handler is being superseded by the one that is [now part of SQLAlchemy](https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#regular-expression-support).   it has a bug in it that will be fixed momentarily.

the behavior you are seeing is because you're applying your event handler at the class level, and the event system applies instance level events after class level events.   apply it to the engine object directly and your event will take place second:

```
engine = sa.create_engine('sqlite://', echo=True)

#@sa.event.listens_for(sa.engine.Engine, 'connect')
@sa.event.listens_for(engine, 'connect')
def sqlite_engine_connect(dbapi_conn, connection_record):
    dbapi_conn.create_function('regexp', 2, _regexp)

```

**Comment by gordthompson** [comment]
Fixed via #5624
